### ISSUE TITLE
Accept JSON fields from "extra" argument

### ISSUE THREAD

**Issue by m-gentil** [googleapis/python-logging#411]
**Is your feature request related to a problem? Please describe.**

When attaching the [CloudLoggingHandler](https://googleapis.dev/python/logging/2.6.0/handlers.html#google.cloud.logging_v2.handlers.handlers.CloudLoggingHandler) to a standard python logger, it is not possible to perform structured logging through the `extra` keyword argument for each logging call. Please refer to the [standard logging python library](https://docs.python.org/3/library/logging.html#logging.Logger.debug) for more documentation about the `extra` parameter. Take for instance the following snippet of code, written in Python 3.9 and using `google-cloud-logging==2.6.0` as a dependency:
```python
import logging
import google.cloud.logging
from google.cloud.logging_v2.handlers import CloudLoggingHandler

client = google.cloud.logging.Client()
handler = CloudLoggingHandler(client)

cloud_logger = logging.getLogger('cloudLogger')
cloud_logger.setLevel(logging.INFO)
cloud_logger.addHandler(handler)

cloud_logger.error('bad news', extra={"error_code": 300})  # Note that we include an extra dictionary
```

The `extra` dictionary is completely lost and does not appear in the received log from the Cloud Logging UI (see screenshot).

 **Describe the solution you'd like**

All properties of a [LogRecord](https://docs.python.org/3/library/logging.html#logging.LogRecord) that are considered as `extra`s appear in the jsonPayload of the related entry in Cloud Logging. This can be implemented for instance in the body of the `emit` method of the CloudLoggingHandler ([reference](https://github.com/googleapis/python-logging/blob/6a837a5d1faab1f9fa8ac94e424e847821a0069f/google/cloud/logging_v2/handlers/handlers.py#L180)).
 
 **Describe alternatives you've considered**

The only alternative solution we found is to use the `log_struct` method of the `google.cloud.logging.logger.Logger()` object directly. This is not a valid solution for us as we make heavy use of the standard logging library shipped with python: using a separate logger would introduce too much overhead for our existing and future projects, while killing the purpose of the centralized logging infrastructure that the python standard logging library provides.

 **Additional context**
![Screenshot 2021-10-04 at 17 46 36](https://user-images.githubusercontent.com/15276897/135888964-1e37501b-c7e9-487c-b003-33499c21adb6.png)

**Comment by daniel-sanche** [comment]
Hey, thanks for filing this

We are adding improved support for `jsonPayload` logs in the upcomming `v3.0.0` release, which is currently [staged in a branch in this repo.](https://github.com/googleapis/python-logging/tree/v3.0.0) With v3, strings that contain parseable json will be rendered as structured json logs in Cloud Logging: `cloud_logger.error(json.dumps(my_dict))`

We also wrote logic in the library to handle dictionary inputs directly: `cloud_logger.error(my_dict)`. But this method is counter to the python-logging spec (which expects string inputs) so JSON strings are recommended

I will leave this open to consider your suggestion of using the extras field for this purpose as well. We are currently using `extras` to [allow over-writing other log entry fields](https://github.com/googleapis/python-logging/blob/db68c16ed70bfa85b6681a8c0dc0977120718cc6/google/cloud/logging_v2/handlers/handlers.py#L90), like trace, span, and labels. We may be able to throw any extra unrecognized `extra` fields in the json payload, but I'll have to think it over. Let me know if you think the method I described above works for your use-case

**Comment by m-gentil** [comment]
Hi Daniel 😄 ,

It's great to see that you are improving this functionality.
Unfortunately I fear that the method that you suggest is not optimal for us. In fact, we use different handlers to route logs to multiple destinations according to each environment (for instance, we stream logs to stdout / stderr when developing locally, while we stream logs to Cloud Logging when running applications in production).

Formatting structured logs as a JSON (`cloud_logger.error(json.dumps(my_dict))`) would render logs unreadable by other handlers (for example this would mean more prolix and less readable output when streaming logs to stdout / stderr or inspecting those messages for debugging purposes).

The only workaround that I can think of on our side would be to develop a custom [Formatter](https://docs.python.org/3/library/logging.html#logging.Formatter) to gather user-defined `extra` fields and formatting the message and its extras as a JSON. Then, we could attach this formatter to the [CloudLoggingHandler](https://googleapis.dev/python/logging/2.6.0/handlers.html#google.cloud.logging_v2.handlers.handlers.CloudLoggingHandler) so that the formatted JSON is later interpreted by the snippet of code you are mentioning.

Still, we would consider this as boilerplate code that we have to keep updated according to future changes in `google-cloud-logging` — this is suboptimal; furthermore we feel like this JSON serialization and successive deserialization introduces overhead that in many cases could be avoided.

That said, do we have any estimate on the release date for 3.0.0?

Please also let us know once you decide whether to go forward with this feature.

**Comment by daniel-sanche** [comment]
That makes sense, thanks. I'll do some research into the best way to do this - either including all unused `extras` in the json payload, or possibly accepting an explicit extra field (something like `cloud_logger.error('bad news', extra={json_fields: my_dict})`)

The hard deadline for v3.0.0 is the end of the year. But I'm planning to do some triage, and maybe bump some things to the next release to try to get it out sooner. I'll try to keep the [milestone](https://github.com/googleapis/python-logging/milestone/1) up to date with my plans

**Comment by stewart-r** [comment]
cool - thanks for responding.

For sure, it's better if that logic is outside the transport layer if that's feasible. Sounds great.

(only reason I had it in there in my proposed solution is because that's where it is now and so took the path of least resistance to surface it in an easily 'overrideable' place)

**Comment by daniel-sanche** [comment]
This will be addressed in the v3.0.0 release
